packages/protocol/contracts/tokenvault/libs/LibVaultUtils.sol

```diff
@@ -0,0 +1,108 @@
+// SPDX-License-Identifier: MIT
+//  _____     _ _         _         _
+// |_   _|_ _(_) |_____  | |   __ _| |__ ___
+//   | |/ _` | | / / _ \ | |__/ _` | '_ (_-<
+//   |_|\__,_|_|_\_\___/ |____\__,_|_.__/__/
+
+pragma solidity ^0.8.20;
+
+import
+    "@openzeppelin/contracts/proxy/transparent/TransparentUpgradeableProxy.sol";
+import { IBridge } from "../../bridge/IBridge.sol";
+import { AddressResolver } from "../../common/AddressResolver.sol";
+
+library LibVaultUtils {
+    /**
+     * Thrown when the sender in a message context is invalid.
+     * This could happen if the sender isn't the expected token vault on the
+     * source chain.
+     */
+    error VAULT_INVALID_SENDER();
+
+    /**
+     * @dev Deploys a contract (via proxy)
+     * @param implementation The new implementation address
+     * @param owner The owner of the proxy admin contract
+     * @param initializationData Data for the initialization
+     */
+    function deployProxy(
+        address implementation,
+        address owner,
+        bytes memory initializationData
+    )
+        external
+        returns (address proxy)
+    {
+        assert(implementation != address(0));
+        proxy = address(
+            new TransparentUpgradeableProxy(implementation, owner, initializationData)
+        );
+    }
+
+    /**
+     * @dev Checks if context is valid
+     * @param validSender The valid sender to be allowed
+     * @param resolver The address of the resolver
+     */
+    function checkValidContext(
+        bytes32 validSender,
+        address resolver
+    )
+        external
+        view
+        returns (IBridge.Context memory ctx)
+    {
+        ctx = IBridge(msg.sender).context();
+        if (
+            ctx.sender
+                != AddressResolver(resolver).resolve(
+                    ctx.srcChainId, validSender, false
+                )
+        ) {
+            revert VAULT_INVALID_SENDER();
+        }
+    }
+
+    /**
+     * @dev Returns the decoded data without selector
+     * @param calldataWithSelector Encoded data with selector
```

Comment:
document return value?

Reply:
To be honest, I still don't know what this function does!
If you just want to remove the selector, why not use https://docs.soliditylang.org/en/v0.6.0/types.html#array-slices